pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go

```diff
@@ -133,22 +129,21 @@ func (r *reconciler) handleMultiArchBuildConfig(ctx context.Context, mabc *v1.Mu
 	}
 
 	targetImageRef := fmt.Sprintf("%s/%s", mabc.Spec.BuildSpec.CommonSpec.Output.To.Namespace, mabc.Spec.BuildSpec.CommonSpec.Output.To.Name)
-	mutateFn := func(mabcToMutate *v1.MultiArchBuildConfig) { mabcToMutate.Status.State = v1.SuccessState }
-
-	if !checkAllBuildsSuccessful(mabc.Status.Builds) {
-		mutateFn = func(mabcToMutate *v1.MultiArchBuildConfig) { mabcToMutate.Status.State = v1.FailureState }
-	} else if err := r.manifestPusher.PushImageWithManifest(mabc.Status.Builds, targetImageRef); err != nil {
-		mutateFn = func(mabcToMutate *v1.MultiArchBuildConfig) {
-			mabcToMutate.Status.Conditions = append(mabcToMutate.Status.Conditions, metav1.Condition{
-				Type:               "PushManifestError",
-				Status:             metav1.ConditionFalse,
-				LastTransitionTime: metav1.Time{Time: time.Now()},
-				Reason:             "PushManifestError",
-				Message:            err.Error(),
-			})
-			mabcToMutate.Status.State = v1.FailureState
-		}
-	}
+	mutateFn := r.runPostBuildActions(
```

Comment:
This fn is just to update the actual resource. All of those operations should be done before that.

Reply:
Yeah all those operations are indeed being done before we apply `mutateFn` . The func `runPostBuildActions` takes care of running them one after the other in a clean way (no unreadable list of `if {} else if {} ...`). `mutateFn` is the final outcome that is going to be applied **after**.

Reply:
The way you are doing it now, its more complicated than the `if else`.

Reply:
This is how it would look like with no refactoring. Other than being more verbose and with some duplications, the worst part is that you have to pass every parameters you need as `registryURL, targetImageRef string` (list will grow as soon as we add more things to do). 
```golang
// runPostBuildActions ensure some actions are performed after the builds have been spawned.
func (r *reconciler) runPostBuildActions2(mabc *v1.MultiArchBuildConfig, registryURL, targetImageRef string) (mutate func(mabcToMutate *v1.MultiArchBuildConfig)) {
	mutateToFailure := func(mabcToMutate *v1.MultiArchBuildConfig) { mabcToMutate.Status.State = v1.FailureState }
	mutateFuncs := make([]func(mabcToMutate *v1.MultiArchBuildConfig), 0)

	defer func() {
		mutate = func(mabcToMutate *v1.MultiArchBuildConfig) {
			for _, f := range mutateFuncs {
				f(mabcToMutate)
			}
		}
	}()

	// Make sure the builds completed successfully
	if !checkAllBuildsSuccessful(mabc.Status.Builds) {
		mutateFuncs = append(mutateFuncs, mutateToFailure)
		return
	}

	// Assemble and push the manifest list
	mutateFn, success := r.handlePushImageWithManifest(mabc, targetImageRef)
	if mutateFn != nil {
		mutateFuncs = append(mutateFuncs, mutateFn)
	}
	if !success {
		mutateFuncs = append(mutateFuncs, mutateToFailure)
		return
	}

	// Mirror to external registries
	srcImage := fmt.Sprintf("%s/%s", registryURL, targetImageRef)
	mutateFn, success = r.handleMirrorImage(srcImage, mabc)
	if mutateFn != nil {
		mutateFuncs = append(mutateFuncs, mutateFn)
	}
	if !success {
		mutateFuncs = append(mutateFuncs, mutateToFailure)
	}
	return
}
```

Reply:
The CR should be updated on every operation. I don't the point of gathering a list of functions and then mutating the object once.  For example, the `checkAllBuildsSuccessful` shouldn't happen inside the postActions. We need to follow the controller's flow. When a build finishes, we update the mabc, this will trigger the reconciliation and we will operate on the object once again. Since now we are adding an extra action, we need to run the manifestPusher first, update the object with the results, and last we should mirror the image to the external registries. This way you keep the flow of the controller. You have to keep in mind that the controller can run in multiple replicas or the single pod can be rescheduled at any time. If you wish to treat the manifest pusher and the mirroring as one step, still the code you shared above, is less complicated than what you tried to introduce in this PR.

Reply:
Definitely I like this approach of one operation at a time. In that case don't we need to make sure an operation has been performed already, like so https://github.com/openshift/ci-tools/pull/3676#discussion_r1371437766 ?